### PR TITLE
[SPARK-33570][SQL][TESTS] Set the proper version of gssapi plugin automatically for MariaDBKrbIntegrationSuite

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR changes mariadb_docker_entrypoint.sh to set the proper version automatically for mariadb-plugin-gssapi-server.
The proper version is based on the one of mariadb-server.
Also, this PR enables to use arbitrary docker image by setting the environment variable `MARIADB_CONTAINER_IMAGE_NAME`.

### Why are the changes needed?

For `MariaDBKrbIntegrationSuite`, the version of `mariadb-plugin-gssapi-server` is currently set to `10.5.5` in `mariadb_docker_entrypoint.sh` but it's no longer available in the official apt repository and `MariaDBKrbIntegrationSuite` doesn't pass for now.
It seems that only the most recent three versions are available for each major version and they are `10.5.6`, `10.5.7` and `10.5.8` for now.
Further, the release cycle of MariaDB seems to be very rapid (1 ~ 2 months) so I don't think it's a good idea to set to an specific version for `mariadb-plugin-gssapi-server`.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Confirmed that `MariaDBKrbIntegrationSuite` passes with the following commands.
```
$  build/sbt -Pdocker-integration-tests -Phive -Phive-thriftserver package "testOnly org.apache.spark.sql.jdbc.MariaDBKrbIntegrationSuite"
```
In this case, we can see what version of `mariadb-plugin-gssapi-server` is going to be installed in the following container log message.
```
Installing mariadb-plugin-gssapi-server=1:10.5.8+maria~focal
```

Or, we can set MARIADB_CONTAINER_IMAGE_NAME for a specific version of MariaDB.
```
$ MARIADB_DOCKER_IMAGE_NAME=mariadb:10.5.6 build/sbt -Pdocker-integration-tests -Phive -Phive-thriftserver package "testOnly org.apache.spark.sql.jdbc.MariaDBKrbIntegrationSuite"
```
```
Installing mariadb-plugin-gssapi-server=1:10.5.6+maria~focal
```

### CODE REVIEW

**Comment by sarutak** [comment]
cc: @dongjoon-hyun @HyukjinKwon @gaborgsomogyi @maropu .

**Comment by sarutak** [comment]
retest this please.

**Comment by gaborgsomogyi** [comment]
I've tried it out but the original test fails as well as this change too:
```
E: Version '1:10.5.5+maria~focal' for 'mariadb-plugin-gssapi-server' was not found
```
Maybe we need to update the `imageName` too.

**Comment by gaborgsomogyi** [comment]
With `mariadb:10.5.8` works fine as you've mentioned in the description.

**Comment by sarutak** [comment]
> I've tried it out but the original test fails as well as this change too:
>
> E: Version '1:10.5.5+maria~focal' for 'mariadb-plugin-gssapi-server' was not found
> 
> Maybe we need to update the imageName too.

It doesn't work with MariaDB 10.5.5 because `mariadb-plugin-gssapi-server:1.10.5.5 is no longer provided by the official apt repository.

> Maybe we need to update the imageName too.

As I mentioned in the description, I don't think it's realistic to hard-code the image name because the release cycle of MariaDB is very fast and only the most recent three versions of plugins are available in the apt repository.

**Review by HyukjinKwon — COMMENTED** [review]
Seems good.

**Comment by maropu** [comment]
> As I mentioned in the description, I don't think it's realistic to hard-code the image name because the release cycle of MariaDB is very fast and only the most recent three versions of plugins are available in the apt repository.

This PR can solve the issue? From the title, I thought this change could make the script automatically fetch the latest image/plugin though.

File: external/docker-integration-tests/src/test/resources/mariadb_docker_entrypoint.sh [hunk 1]
```diff
@@ -18,7 +18,7 @@
 
 dpkg-divert --add /bin/systemctl && ln -sT /bin/true /bin/systemctl
 apt update
-apt install -y mariadb-plugin-gssapi-server=1:10.5.5+maria~focal
+apt install -y mariadb-plugin-gssapi-server=$(dpkg -s mariadb-server | sed -n "s/^Version: \(.*\)/\1/p")
```
Comment (maropu) [inline external/docker-integration-tests/src/test/resources/mariadb_docker_entrypoint.sh:21]: nit: How about logging which version is  installed like this?
```
XXXXX=$(dpkg -s mariadb-server | sed -n "s/^Version: \(.*\)/\1/p")
echo ".... $XXXXX...."
apt install -y mariadb-plugin-gssapi-server=${XXXXXX}
```

**Comment by sarutak** [comment]
> This PR can solve the issue? I read the title and the description and then I thought this change could make the script automatically fetch the latest image/plugin though.

We should use an image where one of the most recent three releases of MariaDB is installed, this PR solves the issue.
If the local image tagged with `10.5` actually refers to one of the non-latest ones (10.5.6 or 10.5.7 for now), the plugin is not for the latest one.

File: external/docker-integration-tests/src/test/resources/mariadb_docker_entrypoint.sh (hunk 1 above)
Comment (sarutak) [reply external/docker-integration-tests/src/test/resources/mariadb_docker_entrypoint.sh:21]: Seems good idea.

**Comment by gaborgsomogyi** [comment]
> As I mentioned in the description, I don't think it's realistic to hard-code the image name because the release cycle of MariaDB is very fast and only the most recent three versions of plugins are available in the apt repository.

OK, then we should fix the PR anyhow because it's constantly failing on my box. Not sure how you've kicked it to passed (`10.5` is not picking up the latest).
Docker logs:
```
$ docker logs 9207ed20d74a
Adding 'local diversion of /bin/systemctl to /bin/systemctl.distrib'
dpkg-divert: warning: please specify --no-rename explicitly, the default will change to --rename in 1.20.x

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Get:1 http://archive.ubuntu.com/ubuntu focal InRelease [265 kB]
Get:2 http://security.ubuntu.com/ubuntu focal-security InRelease [109 kB]
Get:3 http://ftp.osuosl.org/pub/mariadb/repo/10.5/ubuntu focal InRelease [6264 B]
Get:4 http://archive.ubuntu.com/ubuntu focal-updates InRelease [114 kB]
Get:5 http://archive.ubuntu.com/ubuntu focal-backports InRelease [101 kB]
Get:6 http://ftp.osuosl.org/pub/mariadb/repo/10.5/ubuntu focal/main amd64 Packages [17.1 kB]
Get:7 http://archive.ubuntu.com/ubuntu focal/restricted amd64 Packages [33.4 kB]
Get:8 http://archive.ubuntu.com/ubuntu focal/universe amd64 Packages [11.3 MB]
Get:9 http://security.ubuntu.com/ubuntu focal-security/main amd64 Packages [466 kB]
Get:10 http://security.ubuntu.com/ubuntu focal-security/universe amd64 Packages [643 kB]
Get:11 http://security.ubuntu.com/ubuntu focal-security/restricted amd64 Packages [89.4 kB]
Get:12 http://security.ubuntu.com/ubuntu focal-security/multiverse amd64 Packages [1165 B]
Get:13 http://archive.ubuntu.com/ubuntu focal/main amd64 Packages [1275 kB]
Get:14 http://archive.ubuntu.com/ubuntu focal/multiverse amd64 Packages [177 kB]
Get:15 http://archive.ubuntu.com/ubuntu focal-updates/multiverse amd64 Packages [30.4 kB]
Get:16 http://archive.ubuntu.com/ubuntu focal-updates/restricted amd64 Packages [126 kB]
Get:17 http://archive.ubuntu.com/ubuntu focal-updates/universe amd64 Packages [871 kB]
Get:18 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 Packages [850 kB]
Get:19 http://archive.ubuntu.com/ubuntu focal-backports/universe amd64 Packages [4248 B]
Fetched 16.5 MB in 3s (5567 kB/s)
Reading package lists...
Building dependency tree...
Reading state information...
24 packages can be upgraded. Run 'apt list --upgradable' to see them.

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Reading package lists...
Building dependency tree...
Reading state information...
E: Version '1:10.5.5+maria~focal' for 'mariadb-plugin-gssapi-server' was not found
2020-11-27 02:15:16+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 1:10.5.5+maria~focal started.
2020-11-27 02:15:16+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
2020-11-27 02:15:16+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 1:10.5.5+maria~focal started.
2020-11-27 02:15:16+00:00 [Note] [Entrypoint]: Initializing database files
error: Found option without preceding group in config file: /etc/mysql/mariadb.conf.d/auth_gssapi.cnf at line: 1
error: Found option without preceding group in config file: /etc/mysql/mariadb.conf.d/auth_gssapi.cnf at line: 1


PLEASE REMEMBER TO SET A PASSWORD FOR THE MariaDB root USER !
To do so, start the server, then issue the following commands:

'/usr/bin/mysqladmin' -u root password 'new-password'
'/usr/bin/mysqladmin' -u root -h  password 'new-password'

Alternatively you can run:
'/usr/bin/mysql_secure_installation'

which will also give you the option of removing the test
databases and anonymous user created by default.  This is
strongly recommended for production servers.

See the MariaDB Knowledgebase at https://mariadb.com/kb or the
MySQL manual for more instructions.

Please report any problems at https://mariadb.org/jira

The latest information about MariaDB is available at https://mariadb.org/.
You can find additional information about the MySQL part at:
https://dev.mysql.com
Consider joining MariaDB's strong and vibrant community:
https://mariadb.org/get-involved/

2020-11-27 02:15:19+00:00 [Note] [Entrypoint]: Database files initialized
2020-11-27 02:15:19+00:00 [Note] [Entrypoint]: Starting temporary server
2020-11-27 02:15:19+00:00 [Note] [Entrypoint]: Waiting for server startup
error: Found option without preceding group in config file: /etc/mysql/mariadb.conf.d/auth_gssapi.cnf at line: 1
2020-11-27  2:15:19 0 [Note] mysqld (mysqld 10.5.5-MariaDB-1:10.5.5+maria~focal) starting as process 430 ...
```

**Comment by gaborgsomogyi** [comment]
What is the result by others? Have the test passed?

**Comment by sarutak** [comment]
@gaborgsomogyi The version of MariaDB you use seems `10.5.5`.
Could you rm the local image tagged with `10.5` first?

**Comment by gaborgsomogyi** [comment]
Ah, damn docker. Such case I would expect an update, now it works.

**Review by gaborgsomogyi — APPROVED** [review]
LGTM, tested and passed. Thanks!

File: external/docker-integration-tests/src/test/resources/mariadb_docker_entrypoint.sh (hunk 1 above)
Comment (gaborgsomogyi) [reply external/docker-integration-tests/src/test/resources/mariadb_docker_entrypoint.sh:21]: The version can be seen under any circumstances in the log but doesn't hurt to add.
When works fine:
```
Preparing to unpack .../6-mariadb-plugin-gssapi-server_1%3a10.5.8+maria~focal_amd64.deb ...
```
When failed:
```
E: Version '1:10.5.5+maria~focal' for 'mariadb-plugin-gssapi-server' was not found
```

**Comment by maropu** [comment]
> We should use an image where one of the most recent three releases of MariaDB is installed, this PR solves the issue.
If the local image tagged with 10.5 actually refers to one of the non-latest ones (10.5.6 or 10.5.7 for now), the plugin is not for the latest one.

Hm, I see. If it is difficult to follow the rapid release cycle, how about parameterizing the image name as the oracle integration test does so? IMO it would be nicer so that we could always run the test on a latest image:
https://github.com/apache/spark/blob/d082ad0abfe0bc26760626ae0ecb415a8d508a1f/external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala#L63
Then, we can add a document about how-to-run it like `OracleIntegrationSuite`:
https://github.com/apache/spark/blob/d082ad0abfe0bc26760626ae0ecb415a8d508a1f/external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala#L36-L56

**Comment by sarutak** [comment]
@maropu 
Actually, I had the same idea (and reverted...).
https://github.com/apache/spark/pull/30515/commits/48ba4fbdaf993e23a28095277090d4e55ea0bfba

Despite the rapid release cycle, 10.5 series will be maintained until [June 2025](https://mariadb.com/kb/en/changes-improvements-in-mariadb-105/) so I think it's better to refer to `10.5` as the default version.
So, how about the following idea like `DB2IntegrationSuite`?
```
 override val imageName = sys.env.getOrElse("MARIADB_DOCKER_IMAGE_NAME", "10.5") 
```

**Comment by maropu** [comment]
SGTM

File: external/docker-integration-tests/src/test/resources/mariadb_docker_entrypoint.sh (hunk 1 above)
Comment (sarutak) [reply external/docker-integration-tests/src/test/resources/mariadb_docker_entrypoint.sh:21]: Yeah, apt itself logs the version but I wonder if it's better to log what it is trying to install explicitly for clarity.

**Comment by sarutak** [comment]
retest this please.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala [hunk 2]
```diff
@@ -24,15 +24,21 @@ import com.spotify.docker.client.messages.{ContainerConfig, HostConfig}
 import org.apache.spark.sql.execution.datasources.jdbc.connection.SecureConnectionProvider
 import org.apache.spark.tags.DockerTest
 
+/**
+ * To run this test suite for a specific version (e.g., mariadb:10.5.8):
+ * {{{
+ *   MARIADB_DOCKER_IMAGE_NAME=mariadb:10.5.8
+ *     ./build/sbt -Pdocker-integration-tests
+ *     "testOnly org.apache.spark.sql.jdbc.MariaDBKrb2IntegrationSuite"
```
Comment (maropu) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:32]: wrong class name: `MariaDBKrb2IntegrationSuite `

Comment (sarutak) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:32]: Oops. I'll fix it.

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by gaborgsomogyi** [comment]
retest this please

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala [hunk 3]
```diff
@@ -24,15 +24,21 @@ import com.spotify.docker.client.messages.{ContainerConfig, HostConfig}
...
 @DockerTest
 class MariaDBKrbIntegrationSuite extends DockerKrbJDBCIntegrationSuite {
   override protected val userName = s"mariadb/$dockerIp"
   override protected val keytabFileName = "mariadb.keytab"
 
   override val db = new DatabaseOnDocker {
-    // If you change `imageName`, you need to update the version of `mariadb-plugin-gssapi-server`
-    // in `resources/mariadb_docker_entrypoint.sh` accordingly.
-    override val imageName = "mariadb:10.5"
+    override val imageName = sys.env.getOrElse("MARIADB_DOCKER_IMAGE", "mariadb:10.5")
```
Comment (maropu) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:41]: `MARIADB_DOCKER_IMAGE` -> `MARIADB_DOCKER_IMAGE_NAME`

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
The last commit is not related to Jenkins/GA tests, so I merged to master. Thanks!

**Comment by dongjoon-hyun** [comment]
Thank you, @sarutak and @maropu and all!

**Pull request closed without merging**